### ISSUE TITLE
Gzip Compression not working when present in backed configuration

### ISSUE THREAD

**Issue by aliencrash** [jcmoraisjr/haproxy-ingress#848]
**Description of the problem**

with configuration of compression on my backend, there is no compression being used when I use online tools an curl to
check if gzip is being used. 

We are creating the 

**Expected behavior**

It was expected like to get the website application compressed when using tools like https://wtools.io/check-gzip-online.

**Steps to reproduce the problem**

1. Create  a ingress with support of  gzip compression configuration.
2. Try to access the url from https://wtools.io/check-gzip-online with url https://demo-testar-criacao.n360.ao
3. 

**Environment information**

HAProxy Ingress version: `v0.13`

backend n360-saas_demo-testar-criacao-serviceodoo_8069
    mode http
    balance roundrobin
    timeout connect 864000s
    timeout http-keep-alive 864000s
    timeout http-request 864000s
    timeout server 864000s
    acl https-request ssl_fc
    http-request redirect scheme https if !https-request
    http-request set-header X-Original-Forwarded-For %[hdr(x-forwarded-for)] if { hdr(x-forwarded-for) -m found }
    http-request del-header x-forwarded-for
    option forwardfor
    filter compression
    compression algo gzip
    compression type text/css text/less text/plain text/xml application/xml application/json application/javascript
    http-response set-header Strict-Transport-Security "max-age=15768000"
    server srv001 172.16.150.143:8069 weight 1 check inter 2s
    server srv002 127.0.0.1:1023 disabled weight 1 check inter 2s
    server srv003 127.0.0.1:1023 disabled weight 1 check inter 2s
    server srv004 127.0.0.1:1023 disabled weight 1 check inter 2s
    server srv005 127.0.0.1:1023 disabled weight 1 check inter 2s
    server srv006 127.0.0.1:1023 disabled weight 1 check inter 2s
    server srv007 127.0.0.1:1023 disabled weight 1 check inter 2s

**Comment by vladh91** [comment]
I am also seeing same behavior.
@aliencrash just curios how you are creating your ingress? Are you using annotations to enable gzip compression?

**Comment by aliencrash** [comment]
@vladh91 Yes I'm using annotation to create the gzip compression.  I check the haproxy.cfg and everything it's ok in the file. What I posted was actually from that file.

**Comment by jcmoraisjr** [comment]
Hi, this seems to be a HAProxy config issue since the controller doesn't natively support compression. Anyway here is a config attempt and some tests:

1. curl without compression - my ingress has `echoserver.local` as hostname and `3.3.3.4` is the server ip:

```
$ curl -k https://echoserver.local --resolv echoserver.local:443:3.3.3.4 -H 'accept-encoding: gzip'
CLIENT VALUES:
client_address=10.42.0.146
command=GET
...
```

2. adding config-backend annotation to configure compression, here is the ingress after the configuration:

```
$ kubectl get ing echoserver -oyaml |grep -A2 config-backend
    haproxy-ingress.github.io/config-backend: |
      compression algo gzip
      compression type text/css text/less text/plain text/xml application/xml application/json application/javascript
```

3. curl now complains that the output is binary:

```
$ curl -k https://echoserver.local --resolv echoserver.local:443:3.3.3.4 -H 'accept-encoding: gzip'
Warning: Binary output can mess up your terminal. Use "--output -" to tell
Warning: curl to output it to your terminal anyway, or consider "--output
Warning: <FILE>" to save to a file.
```

4. pipe to gunzip to decompress the output:

```
$ curl -sk https://echoserver.local --resolv echoserver.local:443:3.3.3.4 -H 'accept-encoding: gzip' |gunzip
CLIENT VALUES:
client_address=10.42.0.146
command=GET
...
```

5. note that there is no compression without the header, this is by design:

```
$ curl -k https://echoserver.local --resolv echoserver.local:443:3.3.3.4
CLIENT VALUES:
client_address=10.42.0.146
command=GET
...
```

So couldn't reproduce the problem, it seems to work exactly as stated by the HAProxy doc.

**Comment by aliencrash** [comment]
@jcmoraisjr could you share the resulting haproxy.cfg file?

**Comment by jcmoraisjr** [comment]
yep, here is the rendered backend:

```
backend default_echoserver_8080
    mode http
    balance roundrobin
    acl https-request ssl_fc
    http-request set-header X-Original-Forwarded-For %[hdr(x-forwarded-for)] if { hdr(x-forwarded-for) -m found }
    http-request del-header x-forwarded-for
    option forwardfor
    compression algo gzip
    compression type text/css text/less text/plain text/xml application/xml application/json application/javascript
    http-response set-header Strict-Transport-Security "max-age=15768000" if https-request
    server srv001 10.42.0.117:8080 weight 1 check inter 2s
    server srv002 127.0.0.1:1023 disabled weight 1 check inter 2s
    server srv003 127.0.0.1:1023 disabled weight 1 check inter 2s
```

Another behavior that worth to be mentioned is the headers sent by the backend, note eg content-type:

```
$ curl -ik https://echoserver.local --resolv echoserver.local:443:3.3.3.4
HTTP/2 200
server: nginx/1.9.11
date: Mon, 06 Sep 2021 09:04:46 GMT
content-type: text/plain
strict-transport-security: max-age=15768000

CLIENT VALUES:
client_address=10.42.0.146
command=GET
...
```

You should be aware of that, but just to confirm, the [compression keyword doc](http://cbonte.github.io/haproxy-dconv/2.2/configuration.html#4-compression) has a couple of situations where the compression will not work, maybe some of them describes your workload or client.

**Comment by aliencrash** [comment]
Please share with me your echoserver deployment.

**Comment by jcmoraisjr** [comment]
It's the same used in the getting started: https://haproxy-ingress.github.io/docs/getting-started/#deploy-and-expose

**Comment by aliencrash** [comment]
I tried with echo server and  it works. I then try to use it with odoo and nothing is done. Very strange,

I striped all other configurations from the backend to be just like echoserver and nothing. If I do a request the response is coming in http2. from the documentation it's supports http 1.1 and above to enable compression. very strange.

How can I get more information to see whats is happening?

**Comment by aliencrash** [comment]
@jcmoraisjr  when I do  curl -i https://demo-testar-criacao.n360.ao -H 'accept-encoding: gzip'
HTTP/2 200 
content-type: text/html; charset=utf-8
set-cookie: frontend_lang=en_US; Path=/
set-cookie: session_id=a1b0d44afd0ece677febebdf24d130f53a55ae00; Expires=Mon, 06-Dec-2021 13:11:39 GMT; Max-Age=7776000; HttpOnly; Path=/
content-length: 9681
server: Werkzeug/0.14.1 Python/3.7.3
date: Tue, 07 Sep 2021 13:11:39 GMT
strict-transport-security: max-age=15768000
....

**Comment by lemniskett** [comment]
For those who also have issues with gzip compression not working with Odoo, it's because Odoo responses are in HTTP/1.0 in multi-worker mode, not sure why HAProxy's compression doens't work with HTTP/1.0. you'll need to force Odoo to use HTTP/1.1 by changing the `__init__` method in class [odoo.service..server.BaseWSGIServerNoBind](https://github.com/odoo/odoo/blob/1c8e2555366fe6e6b0d74d6db71a29d09bd5f06a/odoo/service/server.py#L106C7-L106C27):
```python
class BaseWSGIServerNoBind(LoggingBaseWSGIServerMixIn, werkzeug.serving.BaseWSGIServer):
    """ werkzeug Base WSGI Server patched to skip socket binding. PreforkServer
    use this class, sets the socket and calls the process_request() manually
    """
    def __init__(self, app):
        handler = werkzeug.serving.WSGIRequestHandler
        handler.protocol_version = "HTTP/1.1"
        werkzeug.serving.BaseWSGIServer.__init__(self, "127.0.0.1", 0, app, handler)
        # Directly close the socket. It will be replaced by WorkerHTTP when processing requests
        if self.socket:
            self.socket.close()
```

Or just use our server-wide module [wsgi_prefork_http11](https://github.com/mplus-oss/cloud-addons/tree/16.0/wsgi_prefork_http11).
